src/vm/checker/typecheck/contexts.rs

```diff
@@ -0,0 +1,149 @@
+use std::collections::{HashMap, BTreeMap};
+use vm::representations::{SymbolicExpression};
+use vm::types::{TypeSignature};
+
+use vm::contexts::MAX_CONTEXT_DEPTH;
+
+use vm::checker::errors::{CheckResult, CheckError, CheckErrors};
+use vm::checker::typecheck::{FunctionType};
+
+const DESERIALIZE_FAIL_MESSAGE: &str = "PANIC: Failed to deserialize bad database data in contract analysis.";
+const SERIALIZE_FAIL_MESSAGE: &str = "PANIC: Failed to deserialize bad database data in contract analysis.";
+
+pub struct TypeMap {
+    map: HashMap<u64, TypeSignature>
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ContractAnalysis {
+    public_function_types: BTreeMap<String, FunctionType>,
+    map_types: BTreeMap<String, (TypeSignature, TypeSignature)>
+}
+
+pub struct TypingContext <'a> {
+    pub variable_types: HashMap<String, TypeSignature>,
+    pub parent: Option<&'a TypingContext<'a>>,
+    pub depth: u16
+}
+
+pub struct ContractContext {
+    pub map_types: HashMap<String, (TypeSignature, TypeSignature)>,
```

Comment:
Since you have getters and setters for these, wanna just make these private?  You rarely access them directly in the code.

Reply:
Yep, fixed, and took the opportunity to add code for the type checker to enforce unique names on those define functions.